### PR TITLE
[Diag] QoI: Suggest fix-it when 'weak' has non-optional type

### PR DESCRIPTION
**Pull request by davezarzycki**
Also:
1) Move and rename a type printer helper function to make this easier.
2) Improve @IBOutlet diagnostic QoI at the same time.

### CODE REVIEW

**Comment by davezarzycki** [comment]
@swift-ci please smoke test

File: include/swift/AST/Types.h
Comment (rudkx) [inline include/swift/AST/Types.h]: I wouldn't use TypeRepr in the name here since this has nothing to do with `TypeRepr`s.

I don't have any great ideas on naming this, though.

File: include/swift/AST/Types.h
Comment (rudkx) [inline include/swift/AST/Types.h]: A comment here would be helpful.

**Review by rudkx — CHANGES_REQUESTED** [review]
Can you add tests for the new fixit?

Otherwise looks good to me.

File: include/swift/AST/Types.h
Comment (davezarzycki) [reply include/swift/AST/Types.h]: I can't think of a better name either. FWIW – If you look at `TypeRepr::isSimple()`, there is very similar logic. The problem is essentially the same and that is why I named the method as such.

If I add cross-referenceing comments to *both* the `TypeRepr` and `TypeBase`, would you be okay with the name?

**Comment by davezarzycki** [comment]
@rudkx – Are we now requiring tests for fixits? I only see a grand total of *six* commits in the last year that added fixit tests. Also, the number of "fixit" test files in `test/FixCode` is seven, and one of those seven is marked XFAIL. :-(

I'm okay with adding the test, but I hate the feeling that I'm being held to a higher standard than the average committer.

File: include/swift/AST/Types.h
Comment (rudkx) [reply include/swift/AST/Types.h]: Sorry I wasn't clear here. This was just a suggestion that the name might be a bit misleading and that a comment on what the method does would be helpful.

I don't have a better suggestion for the name, but I don't think that should hold up merging the change.

**Comment by rudkx** [comment]
@davezarzycki In theory any change that has observable testable effects should have a test to go with it.

In practice sometimes people forget to add tests (and that is sometimes caught in code review), sometimes creating a stand-alone test is impractical (i.e. I've spent hours trying to extract a test case from a large code base and eventually decided it was going to be impractical), and sometimes we just don't have a great way to write direct tests for our test suite (e.g. we don't have a lot of compile-time testing in our regression test suite, just a few scale-tests and a few timer-based test to test expression type checking specifically).

Here are a handful of commits from the last month that I just found that add tests that specifically test for the output from fixits (although they don't necessarily add the fixit they are testing for):

797901fc4145c6437fe9c69edf32d495aefc6723
2b1f7657d4fd44a96e5eb09f6a19f26ab7f2c3e4
6a080d8c5c88b62b3c3f945689b774efd770d808
a9f26ab8930cad34bbab21e2f719bfab413bb22d
da6d078ff945cd66f4479d027ef3cd6ee965808d
7815892a760cfbe5aa6737f8bbde155813c09d7a
ceb51eea80d4741afe13b485b568a56564ea1e06

**Comment by davezarzycki** [comment]
Oh! I see now! I take what I wrote all back. I thought I had to write something like what was in `test/FixCode`. I hadn't paid any attention to the `{{x-y=z}}` test syntax until now. Thanks!

@swift-ci please smoke test

**Review by rudkx — APPROVED** [review]
Thanks Dave, LGTM!

**Pull request merged**
